ksql-common/src/main/java/io/confluent/ksql/function/UdfIndex.java

```diff
@@ -294,30 +291,71 @@ public int hashCode() {
     }
 
     /**
-     * @param argument the argument to test against
+     * @param argument         the argument to test against
+     * @param reservedGenerics a mapping of generics to already reserved types - this map
+     *                         will be updated if the parameter is generic to point to the
+     *                         current argument for future checks to accept
+     *
      * @return whether or not this argument can be used as a value for
      *         this parameter
      */
     // CHECKSTYLE_RULES.OFF: BooleanExpressionComplexity
-    boolean accepts(final Schema argument) {
+    boolean accepts(final Schema argument, final Map<Schema, Schema> reservedGenerics) {
       if (argument == null) {
         return schema.isOptional();
       }
 
-      final Schema.Type type = schema.type();
-      if (!Objects.equals(type, argument.type())) {
-        return false;
+      if (GenericsUtil.hasGenerics(schema)) {
+        return reserveGenerics(schema, argument, reservedGenerics);
       }
 
+      final Schema.Type type = schema.type();
+
       // we require a custom equals method that ignores certain values (e.g.
       // whether or not the schema is optional, and the documentation)
-      return CUSTOM_SCHEMA_EQ.getOrDefault(type, (a, b) -> true).test(schema, argument)
+      return Objects.equals(type, argument.type())
+          && CUSTOM_SCHEMA_EQ.getOrDefault(type, (a, b) -> true).test(schema, argument)
           && Objects.equals(schema.version(), argument.version())
           && Objects.equals(schema.parameters(), argument.parameters())
           && Objects.deepEquals(schema.defaultValue(), argument.defaultValue());
     }
     // CHECKSTYLE_RULES.ON: BooleanExpressionComplexity
 
+    private static boolean reserveGenerics(
+        final Schema schema,
+        final Schema argument,
+        final Map<Schema, Schema> reservedGenerics
+    ) {
+      if (!GenericsUtil.instanceOf(schema, argument)) {
+        return false;
+      }
+
+      final Map<Schema, Schema> genericMapping = GenericsUtil.resolveGenerics(schema, argument);
+      for (final Entry<Schema, Schema> entry : genericMapping.entrySet()) {
```

Comment:
nit: consider pulling this pattern out into a common function in GenericsUtil, e.g:

```
public boolean checkGenericsConsistent(final Collection Entry<Schema, Schema> mappings, Optional<Map<Schema, Schema>> oReserved) {
    final Map<Schema, Schema> reserved = oReserved.orElseGet(new HashMap<>());
    for (final Entry<Schema, Schema> entry : mappings) {
        final Schema old = reserved.putIfAbsent(entry.getKey(), entry.getValue());
        if (old != null && !old.equals(entry.getValue())) {
          return false;
        }
    }
    return true;
}
```

Then you can use it from instanceOf, resolveGenerics, and here

Reply:
I wanted to have as few methods with side-effects as possible, but the problem is that we need to both return a boolean and update a map... 🤔 I'll see if I can make that clean

Reply:
I remember now why I didn't do this originally, for `resolveGenerics` I wanted to be able to have a detailed exception message that tell me exactly which one was the problem. If I throw an exception after returning I lose that information. I think the amount of boiler plate to get this to work for all three of the methods is more confusing than just doing it in place.